mlir/include/mlir/Dialect/MIOpenOps/LowerMIOpenOps.h

```diff
@@ -496,39 +509,80 @@ struct Conv2DRewritePattern : public OpRewritePattern<T> {
 
       llvm::SmallVector<StringAttr, 3> mergedPart1DimNames;
       llvm::SmallVector<IntegerAttr, 3> mergedPart1Dims;
-      // Assume yDim is always less than xDim.
-      if (cDim.getInt() < yDim.getInt()) {
-        mergedPart1DimNames.push_back(cDimName);
-        mergedPart1DimNames.push_back(yDimName);
-        mergedPart1DimNames.push_back(xDimName);
-        mergedPart1Dims.push_back(cDim);
-        mergedPart1Dims.push_back(yDim);
-        mergedPart1Dims.push_back(xDim);
+
+      if (convOpType == miopen::ConvOpType::Conv2DBwdWeightOpType) {
+        // Assume hDim is always less than wDim.
+        if (nDim.getInt() < hDim.getInt()) {
+          mergedPart1DimNames.push_back(nDimName);
+          mergedPart1DimNames.push_back(hDimName);
+          mergedPart1DimNames.push_back(wDimName);
+          mergedPart1Dims.push_back(nDim);
+          mergedPart1Dims.push_back(hDim);
+          mergedPart1Dims.push_back(wDim);
+        } else {
+          mergedPart1DimNames.push_back(hDimName);
+          mergedPart1DimNames.push_back(wDimName);
+          mergedPart1DimNames.push_back(nDimName);
+          mergedPart1Dims.push_back(hDim);
+          mergedPart1Dims.push_back(wDim);
+          mergedPart1Dims.push_back(nDim);
+        }
       } else {
-        mergedPart1DimNames.push_back(yDimName);
-        mergedPart1DimNames.push_back(xDimName);
-        mergedPart1DimNames.push_back(cDimName);
-        mergedPart1Dims.push_back(yDim);
-        mergedPart1Dims.push_back(xDim);
-        mergedPart1Dims.push_back(cDim);
+        // Assume yDim is always less than xDim.
+        if (cDim.getInt() < yDim.getInt()) {
+          mergedPart1DimNames.push_back(cDimName);
+          mergedPart1DimNames.push_back(yDimName);
+          mergedPart1DimNames.push_back(xDimName);
+          mergedPart1Dims.push_back(cDim);
+          mergedPart1Dims.push_back(yDim);
+          mergedPart1Dims.push_back(xDim);
+        } else {
+          mergedPart1DimNames.push_back(yDimName);
+          mergedPart1DimNames.push_back(xDimName);
+          mergedPart1DimNames.push_back(cDimName);
+          mergedPart1Dims.push_back(yDim);
+          mergedPart1Dims.push_back(xDim);
+          mergedPart1Dims.push_back(cDim);
+        }
       }
+
       llvm::SmallVector<StringAttr, 3> mergedPart2DimNames;
       llvm::SmallVector<IntegerAttr, 3> mergedPart2Dims;
-      // Assume hDim is always less than wDim.
-      if (nDim.getInt() < hDim.getInt()) {
-        mergedPart2DimNames.push_back(nDimName);
-        mergedPart2DimNames.push_back(hDimName);
-        mergedPart2DimNames.push_back(wDimName);
-        mergedPart2Dims.push_back(nDim);
-        mergedPart2Dims.push_back(hDim);
-        mergedPart2Dims.push_back(wDim);
+
+      if (convOpType == miopen::ConvOpType::Conv2DBwdWeightOpType) {
+        // Assume yDim is always less than xDim.
```

Comment:
I didn't notice the input transformation since the logic is the same between forward and bwd. Maybe I have some gap in understanding how transformation works. If that's the case feel free to clarify.

You changes (L570-L585) make me wonder if the input layout could be "NCWH"? If that's the case, it will be transformed to "NCXWoYHo", which effectively violated the hDim less than wDim assumption. Looks like we need to handle that situation? 

The other situation is: What if the input layout is "CHNW"? It should be transformed to "CYHoNXWo" to "CYX, HoNWo". I can't see the logic should be able to insert "HoNWo". As if right now, it is either "NHoWo" or "HoWoN".
\+ @whchung

Reply:
The input transformation for wrw is similar as forward: NCHW -> NCXWoYHo, but CYX: GemmN; NWoHo: GemmK

Reply:
Understood. What about other layout mentioned above, such as "NCWH" or "CHNW"? Will those be supported in the input tensor logic here?

Reply:
Technically, MILR should transform input layout to the immediate layout NCXWoYHo, no matter what input layout is

Reply:
Yes, I understand about the target layout to be `NCXWoYHo`. But apparently the comment at line 533 stands for a reason:

> // Assume yDim is always less than xDim.

I'm trying to understand when yDim and xDim position does not satisfy this comment, do we have the correct logic?

Reply:
If I understand correctly, `// Assume yDim is always less than xDim.` means we do not swap yDim and xDim, i.e., KCXY.

Reply:
Let's get clarified on the transformation process first. 

For a input layout of `NCWH`, does it go through the below transformation? If not, what process should it be?

- `NCWH` - `N,C,Hpad,Wpad` - `N,C,Y,Ho,X,Wo` - `CYX,NHoWo`

For a input layout of `CHNW`, does it go through the below transformation? If not, what process should it be?

- `CHNW` - `N,C,Hpad,Wpad` - `N,C,Y,Ho,X,Wo` - `CYX,NHoWo`

Reply:
Yes. No matter what the input layout is: we have intermediate layouts: `N,C,Hpad,Wpad - N,C,Y,Ho,X,Wo`. 

The comments `// Assume yDim is always less than xDim` means we do not support filter layouts that have XDim is less than YDim, i.e., KCXY.

Reply:
Synced offline with jack and confirmed that we only support `YX` & `HW` layout.

For completeness of discussion, it looks like the code below handles different layout such that it treats `HW` as one entity, such that certain other variant of transformation is also allowed.

`NWHC` - `N,Hpad,Wpad,C` - `N,Y,Ho,X,Wo,C` - `YXC,NHoWo`

Reply:
`NHWC - N,Hpad,Wpad,C - N,Y,Ho,X,Wo,C - YXC,NHoWo` technically is possible. But based on currently MLIR implementation, I did not see where we support this.

Reply:
Sounds good. Just want to make sure we catch all scenarios properly.